pkg/querier/http.go

```diff
@@ -502,10 +502,11 @@ func WrapQuerySpanAndTimeout(call string, q *QuerierAPI) middleware.Interface {
 				queryTimeout = q.cfg.QueryTimeout
 			}
 
-			_, cancel := context.WithDeadline(ctx, time.Now().Add(queryTimeout))
+			newCtx, cancel := context.WithTimeout(ctx, queryTimeout)
```

Comment:
This code looks correct to me. Just curious, was it an oversight or intentional to ignore the returned context here in the previous PR?

Reply:
Also, why do we need to do this if the `Engine` can enforce the timeout? As we can see, multiple places to enforce a timeout is causing a lot of trouble.

Reply:
> This code looks correct to me. Just curious, was it an oversight or intentional to ignore the returned context here in the previous PR?

It was for sure an oversight on my side, sorry for that. 

> Also, why do we need to do this if the Engine can enforce the timeout? As we can see, multiple places to enforce a timeout is causing a lot of trouble.

Engine is used for queries and subqueries, while this part of the code is only used by queries.

I agree with you that this mechanism is error prone, but I feel like we can't yet jump to the final solution of getting rid of all timeouts and only relying on an engine timeout.

Reply:
No need to say sorry, it happens and I do mistakes too :)

> I agree with you that this mechanism is error prone, but I feel like we can't yet jump to the final solution of getting rid of all timeouts and only relying on an engine timeout.

Curious, what is stopping us from relying just on the engine? Is it backwards compatibility or confidence in it or something else?

Reply:
I think we can make the engine backwards compatible if we do [this](https://github.com/grafana/loki/pull/7230/files/7ec1b5eaf0597e51d89409253e530283fff82aea#r980698619) unless I am missing something?

Reply:
> I think we can make the engine backwards compatible if we do this unless I am missing something?

agreed, I think we can make it backwards compatible this way.
However, be aware that triggering the timeout here instead of in the engine allow us to also timeout if the request struggle into something before reaching the engine.
On a side note, I would prefer having this PR focused on fixing the current behavior and then in a follow-up PR to centralize timeouts behavior in a single place. Mainly because changing the overall behavior will need some discussion between more maintainers, which will take more time.

Reply:
> However, be aware that triggering the timeout here instead of in the engine allow us to also timeout if the request struggle into something before reaching the engine.

I think timeout would be enforced in the engine to make it work with ruler as well? Anyways, I think the timer should start at the earliest point possible.

> On a side note, I would prefer having this PR focused on fixing the current behavior and then in a follow-up PR to centralize timeouts behavior in a single place. Mainly because changing the overall behavior will need some discussion between more maintainers, which will take more time.

I totally agree. Let us do a separate PR for it.

Reply:
> I think timeout would be enforced in the engine to make it work with ruler as well?

Indeed. 

> Anyways, I think the timer should start at the earliest point possible.

Same! So, I believe this is the main challenge with unifying the timeouts: multiple parts of the code query things in different forms, but we don't want to start counting the deadline too late.

Reply:
Let us give it a try in a separate PR. Going to resolve this for now.